### ISSUE TITLE
Exponential map missing in BaseH?

### ISSUE THREAD

**Issue by gtegner** [HazyResearch/KGEmb#7]
Cool work! 
I noticed a difference with the tensorflow implementation. On line 42 in [1], the get_rhs function which returns the hyperbolic tail embedding maps the output of self.entity by the exponential map to get a point in hyperbolic space, which makes sense. 

However in this implementation (line 36 in [2]), it just returns self.entity. What is the difference here?

I'm also wondering about get_queries in Atth. 

Here, the variable head is the output of self.entity in line 92 
```
head = self.entity(queries[:,0])
```
Shouldn't this also be mapped again by the exponential map to get the hyperbolic embedding, as is specified in equations (8) and (9) in the paper? As:
```
head = expmap0(self.entity(queries[:,0]))
```

As I understand it, the output of the embedding layers give a point in R^d which needs to be mapped to the poincare ball by the exponential map. This seems to be happening in RotH but not in the other classes. 
Is this a bug or am i missing something?

[1] https://github.com/tensorflow/neural-structured-learning/blob/master/research/kg_hyp_emb/models/hyperbolic.py
[2] https://github.com/HazyResearch/KGEmb/blob/master/models/hyperbolic.py

**Comment by ines-chami** [comment]
Good question! The exponential map is actually already part of the `hyp_distance_multi_c` implementation:
https://github.com/HazyResearch/KGEmb/blob/f52b1f83e1aa063e72d9cd4392e6b89cbf85d56a/utils/hyperbolic.py#L152

The reason why the PyTorch and Tensorflow implementations differ is because the Tensorflow implementation does not support multiple curvatures. If we want to support multiple curvatures, we need to map all the candidate tail entities to hyperbolic spaces of different curvatures, based on the relations in a batch of triples. Since this was computationally expensive to implement, we wrote the `hyp_distance_multi_c` which in takes as input:
- hyperbolic queries (output of `get_lhs`) shape B x d
- hyperbolic curvatures shape B x 1
- **euclidean** tail entities shape N x d (in eval mode) and B x d otherwise. 

The `hyp_distance_multi_c` uses the following trick to account for all the different curvatures in eval mode:
![image](https://user-images.githubusercontent.com/13433391/100907436-9dc67800-34ca-11eb-803f-dd6472dc96d0.png)

**Comment by ines-chami** [comment]
On you second point: Rotations and reflections that are centered at the origin are isometries of both the Euclidean and hyperbolic spaces. They commute with the exponential map at the origin and we can therefore apply them before the exponential map (in the tangent space, like in AttH and RefH) or after the exponential map (on hyperbolic embeddings, like in RotH). 

For RotH we found that adding a mobius addition gave better performance, so we needed to apply the exponential mapping before the mobius addition and rotation steps as in:
https://github.com/HazyResearch/KGEmb/blob/f52b1f83e1aa063e72d9cd4392e6b89cbf85d56a/models/hyperbolic.py#L50
